File: src/load_course_input.rs
use std::{
    fs::File,
    io::{BufReader, Error},
    path::PathBuf,
    str::FromStr,
    sync::Arc,
};

use serde_json::{Map, Value};

use crate::{
    course_library::CourseLibrary,
    course_types::{ClassPeriod, ClassTime, Course, CourseDays, CourseSection},
    helpers::err,
};

pub fn load_course_file(path: &PathBuf) -> Result<CourseLibrary, Error> {
    let f = File::open(path)?;
    let reader = BufReader::new(f);
    let json: Value = serde_json::from_reader(reader)?;

    let data_arr = json
        .as_object()
        .and_then(|x| x.get("data"))
        .and_then(|x| x.as_array());

    match data_arr {
        Some(data) => data_array_to_courses(data),
        _ => Err(Error::new(
            std::io::ErrorKind::InvalidData,
            "Bad JSON schema for courses: no 'data' key in top-level object",
        )),
    }
}

fn get_stringy_parsable_key<T: FromStr>(data: &Map<String, Value>, key: &str) -> Result<T, Error> {
    (data
        .get(key)
        .ok_or(err(&("No ".to_owned() + key + " property")))?
        .as_str()
        .ok_or(err(&(key.to_owned() + " property is not a string")))?)
    .parse()
    .map_err(|_| err(&(key.to_owned() + " does not represent a parsable type")))
}

pub fn parse_pipe_sep_list_key<T: FromStr>(
    data: &Map<String, Value>,
    key: &str,
) -> Result<Vec<T>, Error> {
    let string_content = match data
        .get(key)
        .ok_or(err(&("No ".to_owned() + key + " property")))?
        .as_str()
    {
        Some(s) => s,
        None => return Ok(Vec::new()),
    };

    string_content
        .split("|")
        .map(|d| {
            d.parse()
                .map_err(|_| err(&("Could not parse member of ".to_owned() + key + " list: " + d)))
        })
        .collect()
}

pub fn get_string_key(data: &Map<String, Value>, key: &str) -> Result<String, Error> {
    Ok(data
        .get(key)
        .ok_or(err(&("No ".to_owned() + key + " property")))?
        .as_str()
        .unwrap_or("")
        .to_string())
}

fn get_u32_key(data: &Map<String, Value>, key: &str) -> Result<u32, Error> {
    Ok(data
        .get(key)
        .ok_or(err(&("No ".to_owned() + key + " property")))?
        .as_u64()
        .ok_or(err(&(key.to_owned() + " property is not an integer")))? as u32)
}

fn get_yn_bool_key(data: &Map<String, Value>, key: &str) -> Result<bool, Error> {
    Ok(data
        .get(key)
        .ok_or(err(&("No ".to_owned() + key + " property")))?
        .as_str()
        .ok_or(err(&(key.to_owned().clone() + " property is not a string")))?
        == "Y")
}

fn data_array_to_courses(data: &Vec<Value>) -> Result<CourseLibrary, Error> {
    let mut library = CourseLibrary::default();

    for alleged_course in data {
        let course = alleged_course
            .as_object()
            .ok_or(err("data field is not object!"))?;

        let course_name = get_string_key(course, "course")?.into();

        let course_record = library
            .get_by_name(&course_name)
            .and_then(|x| x.upgrade())
            .ok_or(err("Couldn't get Course record from library"))
            .or_else(|_| -> Result<Arc<Course>, Error> {
                Ok(Arc::new(Course {
                    course: course_name.into(),
                    sub_group_list: parse_pipe_sep_list_key(course, "sub_group_list")?,
                    seq_num: get_stringy_parsable_key(course, "seq_num")?,
                    title: get_string_key(course, "title")?,
                    long_title: get_string_key(course, "long_title")?,
                    attr_code: parse_pipe_sep_list_key(course, "attr_code")?.into(),
                    units: get_string_key(course, "units")?,
                }))
            })?;

        library.add_section(Arc::new(CourseSection {
            campus_code: get_stringy_parsable_key(course, "campus_code")?,
            term_code: get_string_key(course, "term_code")?,
            ptrm_code: get_string_key(course, "ptrm_code")?,
            crn: get_stringy_parsable_key(course, "crn")?,
            cap: get_u32_key(course, "cap")?,
            enr: get_u32_key(course, "enr")?,
            permission_only: get_yn_bool_key(course, "permission_only")?,
            instructor: get_string_key(course, "instructor")?,
            schedule: zip_days_and_times(
                parse_pipe_sep_list_key(course, "days")?,
                parse_pipe_sep_list_key(course, "times")?,
            )?,
            location: parse_pipe_sep_list_key(course, "building_abbrev")?,
            note: get_string_key(course, "ssrtext_text")?,
            course: course_record.clone(),
        }));
    }

    Ok(library)
}

fn zip_days_and_times(
    mut days: Vec<CourseDays>,
    mut times: Vec<ClassTime>,
) -> Result<Vec<ClassPeriod>, Error> {
    if days.len() != times.len() {
        return Err(err("Could not turn days/times to schedule"));
    }
    let mut periods: Vec<ClassPeriod> = Vec::new();

    loop {
        let day = match days.pop() {
            None => break,
            Some(day) => day,
        };

        let time = match times.pop() {
            None => break,
            Some(time) => time,
        };

        if let CourseDays::Specified(days) = day {
            if let ClassTime::Specified(time) = time {
                periods.push(ClassPeriod { days, time })
            }
        }
    }

    Ok(periods)
}

File: src/load_course_input.ts
import { ClassPeriod, ClassTime, Course, CourseAttribute, CourseDay, Department, TimeRange } from "./course_types.js"
import { readFileSync } from "node:fs";


export default function load_course_file(path: string): Course[] {
    let f = readFileSync(path).toString();

    let data_arr = JSON.parse(f).data;

    return data_arr.map(data_row_to_course);
}

function get_stringy_number_key(data: any, key: string): number {
    let num = +data[key];
    if (isNaN(num)) throw new Error(`Key ${key} is not a number!`);

    return num;
}

export function parse_pipe_sep_list_key<T>(
    data: any,
    key: string,
    parser: (_: string) => T
): T[] {
    let content = data[key];
    if (!content) return [];

    let pipes = content.split("|");

    return pipes.map(parser);
}

function get_yn_bool_key(data: any, key: string): boolean {
    if (data[key] === "Y") return true;
    else if (data[key] === "N") return false;
    else throw new Error(`Key ${key} is not Y or N; instead, it's ${data[key]}`);
}

function data_row_to_course(data: any): Course {
    return {
        campus_code: get_stringy_number_key(data, "campus_code"),
        term_code: data.term_code,
        ptrm_code: data.ptrm_code,
        crn: get_stringy_number_key(data, "crn"),
        sub_group_list: parse_pipe_sep_list_key(data, "sub_group_list", x => x as Department),
        course: data.course,
        seq_num: get_stringy_number_key(data, "seq_num"),
        title: data.title,
        long_title: data.long_title,
        attr_code: parse_pipe_sep_list_key(data, "attr_code", x => x as CourseAttribute),
        units: data.units,
        cap: data.cap,
        enr: data.enr,
        permission_only: get_yn_bool_key(data, "permission_only"),
        instructor: data.instructor,
        schedule: dayTimeToSchedule(
            parse_pipe_sep_list_key(data, "days", x => x == "-" ? undefined : x.split("") as CourseDay[]),
            parse_pipe_sep_list_key(data, "times", parseTimeRange)
        ),
        location: parse_pipe_sep_list_key(data, "building_abbrev", stripHTML),
        note: data.ssrtext_text,
    };
}

function dayTimeToSchedule(days: (CourseDay[] | undefined)[], times: (TimeRange | undefined)[]): ClassPeriod[] {
    if(days.length != times.length) throw new Error("Could not turn days/times to schedule");
    const periods: ClassPeriod[] = []

    for(let i = days.length; i >= 0; i--) {
        const day = days[i], time = times[i];
        if(day == undefined && time == undefined) continue;

        if(day == undefined || time == undefined) throw new Error("Only one of 'day' and 'time' is undefined: could not combine");

        periods.push({
            days: day,
            time: time
        })
    }

    return periods;
}

function stripHTML(source: string): string {
    let result = "";

    let depth = 0, quoted = false;

    for (const char of source) {
        if (char == '"') quoted = !quoted;
        if (!quoted && char == '<') depth++;
        else if (!quoted && char == '>') depth--;
        else if (!quoted && depth == 0) result += char;
    }

    return result;
}

function parseTimeRange(range: string): TimeRange | undefined {
    if (range == "-") return undefined;

    const [start, end] = range.split("-").map(parseTime);

    return {
        end, start
    }
}

function parseTime(time: string): ClassTime {
    const trimmed = time.trim();

    const [hourStr, minStr] = trimmed.split(":");

    const hour = +hourStr + (trimmed.endsWith("PM") ? 12 : 0);

    const minute = +minStr.substring(0, 2);

    if (isNaN(hour) || isNaN(minute)) throw new Error(`Bad time format ${time}`);

    return {
        hour,
        minute
    }
}